for_dict_challenges_bonus.py

```diff
@@ -66,5 +66,112 @@ def generate_chat_history():
+        self.most_common_uniq_user_seen_by_user_id = {'user_id': 0, 'cnt': 0}  # user_id, с наибольшим числом уникальных просмотров
+        self.num_msg_by_time = {'утром': 0, 'днем': 0, 'вечером': 0}  # распределение сообщений по времени
+        self.most_common_msg_by_time = {'time': '', 'cnt': 0}  # самое популярное время отправки
+
+    def collect_statistics(self, msg: dict):
+        self.msgs[msg['id']] = msg
+        sent_by_user_id = msg['sent_by']
+
+        self.count_by_sender(sent_by_user_id)
+
+        if msg['reply_for']:
+            self.count_by_reply(msg)
+
+        self.count_by_uniq_seen(msg, sent_by_user_id)
+
+        self.count_by_time(msg)
+
+    def count_by_sender(self, sent_by_user_id: int):
+        """
+        Собирает статистику по количеству отправленных сообщений по авторам сообщений и
+        находит айди пользователя, который написал больше всех сообщений
+
+        :param sent_by_user_id:
+        :return:
+        """
+        self.num_msg_by_sender_user_id.setdefault(sent_by_user_id, 0)
+        self.num_msg_by_sender_user_id[sent_by_user_id] += 1
+        if self.num_msg_by_sender_user_id[sent_by_user_id] > self.most_common_sender_user_id['cnt']:
+            self.most_common_sender_user_id['user_id'] = sent_by_user_id
+            self.most_common_sender_user_id['cnt'] = self.num_msg_by_sender_user_id[sent_by_user_id]
+
+    def count_by_reply(self, msg):
+        """
+        Собирает статистику по ответам и находит айди пользователя, на сообщения которого больше всего отвечали
+
+        :param msg:
+        :return:
+        """
+        self.num_msg_reply_by_source_msg_id.setdefault(msg['reply_for'], 0)
+        self.num_msg_reply_by_source_msg_id[msg['reply_for']] += 1
+        user_id_msg_source = self.msgs[msg['reply_for']]['sent_by']
+        self.num_msg_reply_by_source_msg_user_id.setdefault(user_id_msg_source, 0)
+        self.num_msg_reply_by_source_msg_user_id[user_id_msg_source] += 1
+        if self.num_msg_reply_by_source_msg_user_id[user_id_msg_source] > self.most_common_msg_reply_user_id['cnt']:
+            self.most_common_msg_reply_user_id['user_id'] = user_id_msg_source
+            self.most_common_msg_reply_user_id['cnt'] = self.num_msg_reply_by_source_msg_user_id[user_id_msg_source]
+
+    def count_by_uniq_seen(self, msg, sent_by_user_id):
+        """
+        Собирает статистику пользователей, сообщения которых видело больше всего уникальных пользователей
+
+        :param msg:
+        :param sent_by_user_id:
+        :return:
+        """
+        self.uniq_user_seen_by_user_id.setdefault(sent_by_user_id, set())
+        self.uniq_user_seen_by_user_id[sent_by_user_id].update(msg['seen_by'])
+
+    def count_by_time(self, msg: dict):
+        """
+        Собирает статистику по времени и определяет когда в чате больше всего сообщений
+
+        :param msg:
+        :return:
+        """
+        if msg['sent_at'].time() <= msg['sent_at'].time().replace(hour=12, minute=00):
```

Comment:
Что бы упростить весь if, можно обращаться к атрибуту hours:
```suggestion
        if msg['sent_at'].hours <= 12:
```